### DISCUSSION TITLE
Cost function

### DISCUSSION THREAD

**Discussion by ConorWilliams** [ConorWilliams/ballot#3] [General]
Taley wrote via email:

Hi Conor,

Could you provide a clean (mathematical) description of the cost function you're using, in an easily readable format, to make it easier to scrutinise it before the meeting, rather than having to digest the implementation.

Talay

**Comment by ConorWilliams** [comment]
I encourage you to have a look at the [single source file](https://github.com/ConorWilliams/ballot/blob/main/src/cost.hpp) as the cost function is not really mathematically expressible for all the edge cases. That said the most important parts can be elucidated with some graphs.

The segment dealing with assigning a person to a room follows an hyperbolic tangent and ensures the cost is in the range [0,1]. The graph below shows the cost of getting your x^th choice on the interval [0,10):

![2021-02-25-102035_3840x1080_scrot](https://user-images.githubusercontent.com/47435400/109139952-00d74c00-7754-11eb-8aa1-c0465b72efd4.png)

I chose this as I believe we should give more weight to fist/second choices. The cost of assigning someone to a room they don't want is high (~100) to ensure it never happens. Furthermore, there is a penalty for choosing rooms that are not hostels. Any hostel choice is guaranteed to have a lower score than a non-hostel room. This forces the hostels to fill up.

The cost of kicking someone off the ballot depends on their priority level: 

![2021-02-25-105021_3840x1080_scrot](https://user-images.githubusercontent.com/47435400/109142915-56f9be80-7757-11eb-9f3f-34ad7623f295.png)

I wanted to keep the kick-cost close-ish to the cost of getting your last choice hence chose [2,1). This ensures it is normally preferable to get a room. If this is chosen too large then people could potentially exploit the system by adding many honey-pot rooms they don't want to live in to force their first choice.

**Reply by talaych** [reply]
is k the same for everyone, or if you put fewer rooms (if you're allowed to...) does it reflect the number of rooms you put?

**Reply by ConorWilliams** [reply]
Everyone must put the same number of rooms
